File: src/services/userEventService.js
import { collection, doc, serverTimestamp, writeBatch } from 'firebase/firestore';
import { Capacitor } from '@capacitor/core';
import { db } from '../firebase/config';
 
const ALLOWED_EVENT_TYPES = new Set(['impression', 'view', 'scroll', 'like', 'comment', 'share', 'save']);
 
function getDeviceType() {
  try {
    if (Capacitor?.isNativePlatform?.()) {
      const platform = (Capacitor.getPlatform?.() || '').toLowerCase();
      if (platform === 'ios' || platform === 'android') return platform;
      return 'android';
    }
  } catch (_) {}
  return 'web';
}
 
function unixSecondsNow() {
  return Math.floor(Date.now() / 1000);
}
 
function createSessionId() {
  try {
    if (globalThis.crypto?.randomUUID) return globalThis.crypto.randomUUID();
  } catch (_) {}
  return `sess_${Date.now()}_${Math.random().toString(16).slice(2)}`;
}
 
/**
 * User interaction event tracking for feed recommendation signals.
 * - Non-blocking: events are queued and flushed in batches.
 * - Impression dedupe: no duplicate impressions for same post in same session.
 */
class UserEventService {
  constructor() {
    this._sessionId = null;
    this._deviceType = getDeviceType();
    this._queue = [];
    this._flushTimer = null;
    this._impressionDedup = new Set(); // `${sessionId}:${postId}`
    this._maxBatch = 25;
    this._flushIntervalMs = 1200;
  }
 
  startSession() {
    this._sessionId = createSessionId();
    this._impressionDedup.clear();
    return this._sessionId;
  }
 
  getSessionId() {
    return this._sessionId || this.startSession();
  }
 
  /**
   * Enqueue an event. Flushes asynchronously.
   * @param {{ user_id: string, post_id: string, event_type: string, dwell_time_ms?: number|null, position_in_feed?: number|null }} payload
   */
  logEvent(payload) {
    const user_id = (payload?.user_id || '').trim();
    const post_id = (payload?.post_id || '').trim();
    const event_type = (payload?.event_type || '').trim();
 
    if (!user_id || !post_id) return;
    if (!ALLOWED_EVENT_TYPES.has(event_type)) return;
 
    const session_id = this.getSessionId();
 
    // Deduplicate impressions within session
    if (event_type === 'impression') {
      const key = `${session_id}:${post_id}`;
      if (this._impressionDedup.has(key)) return;
      this._impressionDedup.add(key);
    }
 
    const dwell_time_ms =
      event_type === 'view'
        ? Math.max(0, Number(payload?.dwell_time_ms ?? 0))
        : null;
 
    const position_in_feed =
      payload?.position_in_feed == null ? null : Math.max(1, Number(payload.position_in_feed));
 
    const evt = {
      user_id,
      post_id,
      event_type,
      timestamp: unixSecondsNow(),
      dwell_time_ms,
      session_id,
      device_type: this._deviceType,
      position_in_feed
    };
 
    this._queue.push(evt);
    this._scheduleFlush();
  }
 
  _scheduleFlush() {
    if (this._queue.length >= this._maxBatch) {
      this.flush();
      return;
    }
    if (this._flushTimer) return;
    this._flushTimer = setTimeout(() => {
      this._flushTimer = null;
      this.flush();
    }, this._flushIntervalMs);
  }
 
  async flush() {
    if (!this._queue.length) return;
    const events = this._queue.splice(0, this._maxBatch);
 
    // Non-blocking fire-and-forget
    Promise.resolve().then(async () => {
      try {
        const batch = writeBatch(db);
        const colRef = collection(db, 'user_events');
        for (const e of events) {
          const ref = doc(colRef);
          batch.set(ref, { id: ref.id, ...e, createdAt: serverTimestamp() });
        }
        await batch.commit();
      } catch (err) {
        // If commit fails, don't spam; drop events to avoid UI impact.
        console.warn('[user_events] flush failed:', err?.message || err);
      }
    });
  }
}
 
const userEventService = new UserEventService();
 
export { userEventService };
export const USER_EVENT_SCHEMA = {
  collection: 'user_events',
  fields: [
    'id',
    'user_id',
    'post_id',
    'event_type',
    'timestamp',
    'dwell_time_ms',
    'session_id',
    'device_type',
    'position_in_feed'
  ],
  allowed_event_type_values: [...ALLOWED_EVENT_TYPES]
};
 

File: flutter_app/lib/services/user_event_service.dart
import 'dart:async';
import 'dart:io';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import 'package:uuid/uuid.dart';

const _allowedEventTypes = {
  'impression',
  'view',
  'scroll',
  'like',
  'comment',
  'share',
  'save',
};

String _getDeviceType() {
  if (kIsWeb) return 'web';
  if (Platform.isIOS) return 'ios';
  if (Platform.isAndroid) return 'android';
  return 'web';
}

int _unixSecondsNow() => DateTime.now().millisecondsSinceEpoch ~/ 1000;

String _createSessionId() {
  try {
    return const Uuid().v4();
  } catch (_) {
    return 'sess_${DateTime.now().millisecondsSinceEpoch}_${DateTime.now().microsecond.toRadixString(16)}';
  }
}

/// User interaction event tracking for feed recommendation signals.
class UserEventService {
  UserEventService._();

  static final UserEventService instance = UserEventService._();

  factory UserEventService() => instance;

  String? _sessionId;
  final String _deviceType = _getDeviceType();
  final List<Map<String, dynamic>> _queue = [];
  Timer? _flushTimer;
  final Set<String> _impressionDedup = {};
  static const int _maxBatch = 25;
  static const int _flushIntervalMs = 1200;

  String startSession() {
    _sessionId = _createSessionId();
    _impressionDedup.clear();
    return _sessionId!;
  }

  String getSessionId() => _sessionId ?? startSession();

  void logEvent(Map<String, dynamic> payload) {
    final userId = (payload['user_id'] as String? ?? '').trim();
    final postId = (payload['post_id'] as String? ?? '').trim();
    final eventType = (payload['event_type'] as String? ?? '').trim();

    if (userId.isEmpty || postId.isEmpty) return;
    if (!_allowedEventTypes.contains(eventType)) return;

    final sessionId = getSessionId();

    if (eventType == 'impression') {
      final key = '$sessionId:$postId';
      if (_impressionDedup.contains(key)) return;
      _impressionDedup.add(key);
    }

    final dwellTimeMs = eventType == 'view'
        ? ((payload['dwell_time_ms'] as num?) ?? 0).toInt().clamp(0, 999999999)
        : null;

    final positionInFeed = payload['position_in_feed'] == null
        ? null
        : ((payload['position_in_feed'] as num).toInt()).clamp(1, 999999);

    _queue.add({
      'user_id': userId,
      'post_id': postId,
      'event_type': eventType,
      'timestamp': _unixSecondsNow(),
      'dwell_time_ms': dwellTimeMs,
      'session_id': sessionId,
      'device_type': _deviceType,
      'position_in_feed': positionInFeed,
    });

    _scheduleFlush();
  }

  void _scheduleFlush() {
    if (_queue.length >= _maxBatch) {
      flush();
      return;
    }
    _flushTimer ??= Timer(
      const Duration(milliseconds: _flushIntervalMs),
      () {
        _flushTimer = null;
        flush();
      },
    );
  }

  Future<void> flush() async {
    if (_queue.isEmpty) return;
    final events = _queue.take(_maxBatch).toList();
    _queue.removeRange(0, events.length);

    unawaited(() async {
      try {
        final batch = FirebaseFirestore.instance.batch();
        final colRef = FirebaseFirestore.instance.collection('user_events');
        for (final e in events) {
          final ref = colRef.doc();
          batch.set(ref, {
            'id': ref.id,
            ...e,
            'createdAt': FieldValue.serverTimestamp(),
          });
        }
        await batch.commit();
      } catch (err) {
        debugPrint('[user_events] flush failed: $err');
      }
    }());
  }
}

final userEventService = UserEventService.instance;

const userEventSchema = {
  'collection': 'user_events',
  'fields': [
    'id',
    'user_id',
    'post_id',
    'event_type',
    'timestamp',
    'dwell_time_ms',
    'session_id',
    'device_type',
    'position_in_feed',
  ],
  'allowed_event_type_values': _allowedEventTypes,
};
